Add documentation for shoot access (#5849)
gardener/gardener: docs/usage/shoot_access.md

# Accessing Shoot Clusters

> After creation of a shoot clusters, end-users requires a `kubeconfig` to access it. There are several options available to get to such `kubeconfig`.

Reviewer: ```suggestion
After creation of a shoot clusters, end-users require a `kubeconfig` to access it. There are several options available to get to such `kubeconfig`.
```

> ## Static Token Kubeconfig :

Reviewer: ```suggestion
## Static Token Kubeconfig
```

This `kubeconfig` contains a [static token](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#static-token-file) and provides `cluster-admin` privileges.
It is created by default and persisted in the `<shoot-name>.kubeconfig` secret in the project namespace in the garden cluster.

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
...
spec:

> kubernetes:

Reviewer: ```suggestion
  kubernetes:
```

    enableStaticTokenKubeconfig: true
...
```

> It is **not** the recommended method to access the shoot cluster as the static token `kubeconfig` has some security flaws associated with it:

Reviewer: ```suggestion

It is **not** the recommended method for accessing the shoot cluster as the static token `kubeconfig` has some security flaws associated with it:
```

> - The static token in the `kubeconfig` doesn't have any expiration date. To revoke the static token, the user needs to rotate the kuebcconfig credentials (see https://github.com/gardener/gardener/blob/master/docs/usage/shoot_operations.md#rotate-kubeconfig-credentials).

Reviewer: ```suggestion
- The static token in the `kubeconfig` doesn't have any expiration date. To revoke the static token, the user needs to rotate the kubeconfig credentials (see https://github.com/gardener/gardener/blob/master/docs/usage/shoot_operations.md#rotate-kubeconfig-credentials).
```

- The static token doesn't have any user identity associated with it. The user in that token will always be `system:cluster-admin` irrespective of the person accessing the cluster. Hence, it is impossible to audit the events in cluster.

## `shoots/adminkubeconfig` subresource

> The [`shoots/adminkubeconfig`](https://github.com/gardener/gardener/blob/master/docs/proposals/16-adminkubeconfig-subresource.md) subresource allows users to dynamically generate short-lived `kubeconfig` that can be used to access shoot cluster with `cluster-admin` privileges. The credentials associated with this `kubeconfig` are client certificates which have a very short validity and must be renewed before they expire.

Reviewer: ```suggestion
The [`shoots/adminkubeconfig`](https://github.com/gardener/gardener/blob/master/docs/proposals/16-adminkubeconfig-subresource.md) subresource allows users to dynamically generate temporary `kubeconfig`s that can be used to access shoot cluster with `cluster-admin` privileges. The credentials associated with this `kubeconfig` are client certificates which have a very short validity and must be renewed before they expire (by calling the subresource  endpoint again).
```

The username associated with such `kubeconfig` will be the same which is used for authenticating to the Gardener API. Apart from this advantage, the created `kubeconfig` will not be persisted anywhere.

> 

Reviewer: ```suggestion

In order to request such a `kubeconfig`, you can run the following commands:

```

```bash
export NAMESPACE=my-namespace
export SHOOT_NAME=my-shoot
kubectl create \
    -f <path>/<to>/kubeconfig-request.json \
    --raw /apis/core.gardener.cloud/v1beta1/namespaces/${NAMESPACE}/shoots/${SHOOT_NAME}/adminkubeconfig | jq -r ".status.kubeconfig" | base64 -d
```

Here, the `kubeconfig-request.json` has the following content:

```json
{
    "apiVersion": "authentication.gardener.cloud/v1alpha1",
    "kind": "AdminKubeconfigRequest",
    "spec": {
        "expirationSeconds": 1000
    }
}
```

> > The [`gardenctl-v2`](https://github.com/gardener/gardenctl-v2/) tool makes it easy to target shoot clusters and automatically renews such `kubeconfig` when required.

Reviewer: ```suggestion

> The [`gardenctl-v2`](https://github.com/gardener/gardenctl-v2/) tool makes it easy to target shoot clusters and automatically renews such `kubeconfig` when required.
```

> ## `oidcConfig`

Reviewer: ```suggestion
## OpenID Connect
```

> Gardener has `ClusterOpenIDConnectPreset` and `OpenIDConnectPreset` API resource for injecting additional runtime [OIDC](https://openid.net/connect/) requirements into a Shoot at creation time. The injected information contains configuration for the Kube API Server and optionally configuration for kubeconfig generation using said configuration. (Cluster)OpenIDConnectPreset contains the oidc server details like `clientID` , `issuerURL`. ClusterOpenIDConnectPreset specified oidc configuration applies to `Projects` and `Shoots` cluster-wide while OpenIDConnectPreset is `Project` scoped.
> 
> Gardener provides an admission controller (ClusterOpenIDConnectPreset and OpenIDConnectPreset) which, when enabled, applies ClusterOpenIDConnectPresets and OpenIDConnectPresets respectivly to incoming `Shoot` creation requests having label `oidc : enable` set. It mutates the shoot's `.spec.kubernetes.kubeAPIServer.oidcConfig` field with the oidc configuration which will be used for  `kubeconfig` generation.
> 
> It is the end-user responsibility to incorporate the OpenID Connect configurations in kubeconfig to access the cluster. One way is to use the kubectl plugin called [`kubectl oidc-login`](https://github.com/int128/kubelogin)  for OIDC authentication.
> 
> For further information on (Cluster)OpenIDConnectPreset, refer to [doc](https://github.com/gardener/gardener/blob/master/docs/usage/openidconnect-presets.md).

Reviewer: ```suggestion
The `kube-apiserver` of shoot clusters can be provided with [OpenID Connect configuration](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#openid-connect-tokens) via the `ShootSpec`:

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
...
spec:
  kubernetes:
    oidcConfig:
      ...
\```

It is the end-user's responsibility to incorporate the OpenID Connect configurations in `kubeconfig` for accessing the cluster (i.e., Gardener will not automatically generate `kubeconfig` based on these OIDC settings).
The recommended way is using the `kubectl` plugin called [`kubectl oidc-login`](https://github.com/int128/kubelogin) for OIDC authentication.

If you want to use the same OIDC configuration for all your shoots by default then you can use the `ClusterOpenIDConnectPreset` and `OpenIDConnectPreset` API resources. They allowing defaulting the `.spec.kubernetes.kubeAPIServer.oidcConfig` fields for newly created `Shoot`s such that you don't have to repeat yourself every time (similar to `PodPreset` resources in Kubernetes).
`ClusterOpenIDConnectPreset` specified OIDC configuration applies to `Projects` and `Shoots` cluster-wide (hence, only available to Gardener operators) while `OpenIDConnectPreset` is `Project`-scoped.
Shoots have to "opt-in" for such defaulting by using the `oidc=enable` label.

For further information on `(Cluster)OpenIDConnectPreset`, refer to [this document](https://github.com/gardener/gardener/blob/master/docs/usage/openidconnect-presets.md).
```

> # Shoot_access.md

Reviewer: ```suggestion
# Accessing Shoot Clusters
```

Please also add this document to the index (`docs/README.md`).

> There are several scenarios where end-user would like to access shoot cluster. There are several options available to user to access the shoot cluster.

Reviewer: ```suggestion
After creation of a shoot clusters, end-users require a `kubeconfig` to access it. There are several options available to get to such `kubeconfig`.
```

> ## Static token kubeconfig :

Reviewer: ```suggestion
## Static Token Kubeconfig
```

>   - To access the shoot cluster with the static token kubeconfig, shoot should have `.spec.kubernetes.enableStaticTokenKubeconfig` option set to `true`. enableStaticTokenKubeconfig will allow the creation of kubeconfig secret in the project namespace. End users can fetch the secret and use the kubeconfig inside it. Static token belongs to `system:masters` group and grants `cluster-admin` privileges to cluster.

Reviewer: ```suggestion
This `kubeconfig` contains a [static token](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#static-token-file) and provides `cluster-admin` privileges.
It is created by default and persisted in the `<shoot-name>.kubeconfig` secret in the project namespace in the garden cluster.
```

>     ```yaml
>     apiVersion: core.gardener.cloud/v1beta1
>     kind: Shoot
>     metadata:
>       name: local
>       namespace: garden-local
>       ...
>     spec:
>       kubernetes:
>         version: 1.23.1
>         enableStaticTokenKubeconfig: true
>       ...
>     ```
>   This is not the recommended method to access the shoot cluster as the static token kubeconfig has some security flaws associated with it.
> 
>    - Static token in the kubeconfig doesn't have any expiration date. To revoke the static token, the user needs to rotate the kuebcconfig credentials.
>    - Static token doesn't have user identity associated with it. User in that token will always be system:cluster-admin irrespective of the person accessing the cluster. Hence it is impossible to audit the events in cluster.

Reviewer: ```suggestion
```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
...
spec:
  kubernetes:
    enableStaticTokenKubeconfig: true
...
\```

It is **not** the recommended method to access the shoot cluster as the static token `kubeconfig` has some security flaws associated with it:

- The static token in the `kubeconfig` doesn't have any expiration date. To revoke the static token, the user needs to rotate the kuebcconfig credentials (see https://github.com/gardener/gardener/blob/master/docs/usage/shoot_operations.md#rotate-kubeconfig-credentials).
- The static token doesn't have any user identity associated with it. The user in that token will always be `system:cluster-admin` irrespective of the person accessing the cluster. Hence, it is impossible to audit the events in cluster.
```

>   To disable the creation of static token kubeconfig, set the value of `.spec.kubernetes.enableStaticTokenKubeconfig`  field in the shoot spec to `false`.
>   ```yaml
>     apiVersion: core.gardener.cloud/v1beta1
>     kind: Shoot
>     metadata:
>       name: local
>       namespace: garden-local
>       ...
>     spec:
>       kubernetes:
>         version: 1.23.1
>         enableStaticTokenKubeconfig: true
>       ...
>   ```

Reviewer: ```suggestion
```

> ## Dynamic kubeconfig

Reviewer: ```suggestion
## `shoots/adminkubeconfig` subresource
```

>   Shoot subresource called `adminKubeconfig` allows user to dynamically generate short lived `kubeconfig` that can be used to access shoot cluster with `cluster-admin` priviledge. `shoots/adminkubeconfig` resource can only accept `CREATE` calls and accept   `AdminKubeconfigRequest` .Kubeconfig is generated only when `AdminKubeconfigRequest` sent to the subresource.

Reviewer: ```suggestion
The `shoots/adminkubeconfig` subresource allows users to dynamically generate short-lived `kubeconfig`s that can be used to access shoot cluster with `cluster-admin` privileges. The credentials associated with this `kubeconfig` are client certificates which have a very short validity and must be renewed before they expire.
```

>   User identity of kubeconfig will be the user authenticating to gardener API server. Following command can be used to request a kubeconfig using the `AdminKubeconfigRequest`.

Reviewer: ```suggestion
The username associated with such `kubeconfig` will be the same which is used for authenticating to the Gardener API. Apart from this advantage, the created `kubeconfig`s will not be persisted anywhere.

The following command can be used to request a `kubeconfig` using the `AdminKubeconfigRequest` resource.
```

>   ```bash
>   export NAMESPACE=my-namespace
>   export SHOOT_NAME=my-shoot
>   kubectl create \
>       -f example/shoot-kubeconfig/kubeconfig-request.json \
>       --raw /apis/core.gardener.cloud/v1beta1/namespaces/${NAMESPACE}/shoots/${SHOOT_NAME}/adminkubeconfig | jq -r ".status.kubeconfig" | base64 -d
>   ```
> 
>   Here  `kubeconfig-request.json` has the followng content where `expirationSeconds` is in seconds.
> 
>   ```json
>   {
>       "apiVersion": "authentication.gardener.cloud/v1alpha1",
>       "kind": "AdminKubeconfigRequest",
>       "spec": {
>           "expirationSeconds": 1000
>       }
>   }
>   ```
> 

Reviewer: ```suggestion
```bash
export NAMESPACE=my-namespace
export SHOOT_NAME=my-shoot
kubectl create \
    -f example/shoot-kubeconfig/kubeconfig-request.json \
    --raw /apis/core.gardener.cloud/v1beta1/namespaces/${NAMESPACE}/shoots/${SHOOT_NAME}/adminkubeconfig | jq -r ".status.kubeconfig" | base64 -d
\```

Here, the `kubeconfig-request.json` has the following content:

```json
{
    "apiVersion": "authentication.gardener.cloud/v1alpha1",
    "kind": "AdminKubeconfigRequest",
    "spec": {
        "expirationSeconds": 1000
    }
}
\```
```

> 
>   

Reviewer: ```suggestion
> The [`gardenctl-v2`](https://github.com/gardener/gardenctl-v2/) tool makes it easy to target shoot clusters and automatically renews such `kubeconfig`s when required.
```

# Accessing Shoot Clusters

> After creation of a shoot clusters, end-users require a `kubeconfig` to access it. There are several options available to get to such `kubeconfig`.

Reviewer 2: ```suggestion
After creation of a shoot cluster, end-users require a `kubeconfig` to access it. There are several options available to get to such `kubeconfig`.
```

## Static Token Kubeconfig

This `kubeconfig` contains a [static token](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#static-token-file) and provides `cluster-admin` privileges.
It is created by default and persisted in the `<shoot-name>.kubeconfig` secret in the project namespace in the garden cluster.

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
...
spec:
  kubernetes:
    enableStaticTokenKubeconfig: true
...
```

It is **not** the recommended method to access the shoot cluster as the static token `kubeconfig` has some security flaws associated with it:
- The static token in the `kubeconfig` doesn't have any expiration date. To revoke the static token, the user needs to rotate the kubeconfig credentials (see [rotate-kubeconfig-credentials](https://github.com/gardener/gardener/blob/master/docs/usage/shoot_operations.md#rotate-kubeconfig-credentials)).
- The static token doesn't have any user identity associated with it. The user in that token will always be `system:cluster-admin` irrespective of the person accessing the cluster. Hence, it is impossible to audit the events in cluster.

## `shoots/adminkubeconfig` subresource

> The [`shoots/adminkubeconfig`](https://github.com/gardener/gardener/blob/master/docs/proposals/16-adminkubeconfig-subresource.md) subresource allows users to dynamically generate temporary `kubeconfig`s that can be used to access shoot cluster with `cluster-admin` privileges. The credentials associated with this `kubeconfig` are client certificates which have a very short validity and must be renewed before they expire (by calling the subresource  endpoint again).

Reviewer 2: ```suggestion
The [`shoots/adminkubeconfig`](https://github.com/gardener/gardener/blob/master/docs/proposals/16-adminkubeconfig-subresource.md) subresource allows users to dynamically generate temporary `kubeconfig`s that can be used to access shoot cluster with `cluster-admin` privileges. The credentials associated with this `kubeconfig` are client certificates which have a very short validity and must be renewed before they expire (by calling the subresource endpoint again).
```

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
...
spec:
  kubernetes:
    oidcConfig:
      ...
```

It is the end-user's responsibility to incorporate the OpenID Connect configurations in `kubeconfig` for accessing the cluster (i.e., Gardener will not automatically generate `kubeconfig` based on these OIDC settings).
The recommended way is using the `kubectl` plugin called [`kubectl oidc-login`](https://github.com/int128/kubelogin) for OIDC authentication.

If you want to use the same OIDC configuration for all your shoots by default then you can use the `ClusterOpenIDConnectPreset` and `OpenIDConnectPreset` API resources. They allow defaulting the `.spec.kubernetes.kubeAPIServer.oidcConfig` fields for newly created `Shoot`s such that you don't have to repeat yourself every time (similar to `PodPreset` resources in Kubernetes).
`ClusterOpenIDConnectPreset` specified OIDC configuration applies to `Projects` and `Shoots` cluster-wide (hence, only available to Gardener operators) while `OpenIDConnectPreset` is `Project`-scoped.
Shoots have to "opt-in" for such defaulting by using the `oidc=enable` label.

> For further information on `(Cluster)OpenIDConnectPreset`, refer to [this document](https://github.com/gardener/gardener/blob/master/docs/usage/openidconnect-presets.md).

Reviewer 2: I guess the relative link is the preferred way. Similar changes in other places.
```suggestion
For further information on `(Cluster)OpenIDConnectPreset`, refer to [this document](openidconnect-presets.md).
```

Reviewer 2: /cc @rfranzke
